src/Tools/ExternalAccess/Debugger/GlassTestsHotReloadService.cs

```diff
@@ -0,0 +1,115 @@
+        {
+            _encService = services.GetRequiredService<IEditAndContinueWorkspaceService>();
+            _debuggerService = debuggerService;
+        }
+
+        public async Task StartSessionAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var newSessionId = await _encService.StartDebuggingSessionAsync(
+                solution,
+                new ManagedHotReloadServiceImpl(_debuggerService),
+                captureMatchingDocuments: ImmutableArray<DocumentId>.Empty,
+                captureAllMatchingDocuments: true,
+                reportDiagnostics: false,
+                cancellationToken).ConfigureAwait(false);
+
+            Contract.ThrowIfFalse(_sessionId == default, "Session already started");
+            _sessionId = newSessionId;
+        }
+
+        private DebuggingSessionId GetSessionId()
+        {
+            var sessionId = _sessionId;
+            Contract.ThrowIfFalse(sessionId != default, "Session has not started");
+
+            return sessionId;
+        }
+
+        public void EnterBreakState()
+        {
+            _encService.BreakStateOrCapabilitiesChanged(GetSessionId(), inBreakState: true, out _);
+        }
+
+        public void ExitBreakState()
+        {
+            _encService.BreakStateOrCapabilitiesChanged(GetSessionId(), inBreakState: false, out _);
+        }
+
+        public void OnCapabilitiesChanged()
+        {
+            _encService.BreakStateOrCapabilitiesChanged(GetSessionId(), inBreakState: null, out _);
+        }
+
+        public void CommitSolutionUpdate()
+        {
+            _encService.CommitSolutionUpdate(GetSessionId(), out _);
+        }
+
+        public void DiscardSolutionUpdate()
+        {
+            _encService.DiscardSolutionUpdate(GetSessionId());
+        }
+
+        public void EndDebuggingSession()
+        {
+            _encService.EndDebuggingSession(GetSessionId(), out _);
+            _sessionId = default;
+        }
+
+        public async ValueTask<bool> HasChangesAsync(Solution solution, string? sourceFilePath, CancellationToken cancellationToken)
+        {
+            var sessionId = _sessionId;
+            if (sessionId == default)
+            {
+                return false;
+            }
+
+            return await _encService.HasChangesAsync(sessionId, solution, s_noActiveStatementSpanProvider, sourceFilePath, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async ValueTask<ManagedModuleUpdates> GetEditAndContinueUpdatesAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var result = await GetDebuggingSession().EmitSolutionUpdateAsync(solution, s_noActiveStatementSpanProvider, cancellationToken).ConfigureAwait(false);
+
+            return result.ModuleUpdates.FromContract();
+        }
+
+        public async ValueTask<ManagedHotReloadUpdates> EmitSolutionUpdateAsync(Solution solution, CancellationToken cancellationToken)
```

Comment:
nit:
```suggestion
        public async ValueTask<ManagedHotReloadUpdates> GetHotReloadUpdatesAsync(Solution solution, CancellationToken cancellationToken)
```

so it's consistent with IManagedHotReloadLanguageService